src/levanter/models/whisper.py

```diff
@@ -0,0 +1,572 @@
+    Vocab: Axis = eqx.static_field()
+    config: WhisperConfig = eqx.static_field()
+
+    token_embeddings: hnn.Embedding
+    position_embeddings: hnn.Embedding
+
+    @staticmethod
+    def init(Vocab: Axis, config: WhisperConfig, *, key) -> "WhisperDecoderEmbeddings":
+        k_wte, k_wpe, k_out = haliax.jax_utils.maybe_rng_split(key, 3)
+
+        token_embeddings = hnn.Embedding.init(
+            Vocab, config.Embed, key=k_wte, initializer_range=config.initializer_range
+        )
+
+        # Whisper Initializes the Positional Embeddings as Empty
+        position_embeddings = hnn.Embedding.init(config.Pos, config.Embed, key=k_wpe, initializer_range=0)
+
+        return WhisperDecoderEmbeddings(Vocab, config, token_embeddings, position_embeddings)
+
+    @named_call
+    def embed(self, input_ids, *, key):
+        input_embeds = self.token_embeddings(input_ids)
+        input_Pos = input_ids.resolve_axis("position")
+        position_embeds = self.position_embeddings.embed(hax.arange(input_Pos))
+        x = input_embeds + position_embeds
+
+        return x
+
+    def unembed(self, x: NamedArray):
+        return hax.dot("embed_dim", x, self.token_embeddings.weight)
+
+    def resize_embeddings(self, new_size: int, key: Optional[PRNGKeyArray] = None):
+        new_token_embeddings = self.token_embeddings.resize_embeddings(new_size, key=key)
+        return dataclasses.replace(self, Vocab=self.Vocab.resize(new_size), token_embeddings=new_token_embeddings)
+
+    def _state_dict_key_map(self) -> Dict[str, Optional[str]]:
+        return {"token_embeddings": "embed_tokens", "position_embeddings": "embed_positions"}
+
+
+class WhisperDecoder(eqx.Module, StateDictSerializationMixin):
+    transformer: WhisperTransformer
+    embeddings: WhisperDecoderEmbeddings
+
+    @property
+    def config(self):
+        return self.embeddings.config
+
+    @property
+    def Vocab(self) -> Axis:
+        return self.embeddings.Vocab
+
+    @property
+    def Pos(self) -> Axis:
+        return self.config.Pos
+
+    @classmethod
+    def init(cls, config: WhisperConfig, *, key) -> "WhisperDecoder":
+        k_t, k_embeddings = haliax.jax_utils.maybe_rng_split(key, 2)
+        transformer = WhisperTransformer.init(
+            config.DecoderLayer,
+            config.DecoderHeads,
+            config.DecoderHeadSize,
+            config.DecoderMlp,
+            config,
+            has_cross=True,
+            key=k_t,
+        )
+        embeddings = WhisperDecoderEmbeddings.init(config.Vocab, config, key=k_embeddings)
+
+        return WhisperDecoder(transformer, embeddings)
+
+    def __call__(
+        self,
+        input_ids: NamedArray,
+        audio_embeds: NamedArray,
+        attn_mask: Optional[AttentionMask | NamedArray] = None,
+        *,
+        key=None,
+    ) -> NamedArray:
+        Pos = input_ids.resolve_axis("position")
+        causal_mask = hax.nn.attention.causal_mask(Pos, Pos.alias("key_position"))
```

Comment:
sorry another suggestion based on changes in @vadam5 's branch: can you use AttentionMask.causal? It'll play better with TransformerEngine's flash attention thing... I'm happy to merge as is, it'll just future proof a bit better